### PR TITLE
Process LoadPath directories in sorted order

### PR DESCRIPTION
**Pull request by jfehrle**
Currently Coq adds items from a directory by stepping through the directory in the order the items appear in the directory, which is nondeterministic and not user friendly (i.e. the LoadPath items for a given -Q or -R are not in alphabetic order).  (Recall that `ls` defaults to sorting the entries.)

I think it's unlikely (perhaps impossible) this reordering will impact users.  Given that the order is currently nondeterministic, this seems like a reasonable thing to do.

- [x] Added **changelog**.

<details><summary>For example:</summary>

**Current behavior**
Unsorted directory order:

```
$ ls -U _build/default/theories
ssrmatching  Unicode      Reals        Sets         funind         NArith
Bool         Wellfounded  btauto       Init         Structures     rtauto
Program      Classes      omega        extraction   Array          Relations
Numbers      Lists        PArith       _CoqProject  Sorting        QArith
Arith        Strings      ssr          Setoids      dune.disabled  Logic
MSets        dune         index.mld    Vectors      FSets          derive
Floats       ZArith       setoid_ring  micromega    nsatz          Compat
```

Print LoadPath.
(* notice that the subdirectories for Coq.Numbers are in scrambled order, too *)

Logical Path / Physical path:
<> /home/jfehrle/.local/share/coq
<> /home/proj/coq/_build/default/user-contrib
Ltac2 /home/proj/coq/_build/default/user-contrib/Ltac2
Coq /home/proj/coq/_build/default/theories
Coq.**ssrmatching** /home/proj/coq/_build/default/theories/ssrmatching
Coq.**Bool** /home/proj/coq/_build/default/theories/Bool
Coq.**Program** /home/proj/coq/_build/default/theories/Program
Coq.**Numbers** /home/proj/coq/_build/default/theories/Numbers
Coq.Numbers.**Natural** /home/proj/coq/_build/default/theories/Numbers/Natural
Coq.Numbers.Natural.Abstract
  /home/proj/coq/_build/default/theories/Numbers/Natural/Abstract
Coq.Numbers.Natural.Peano
  /home/proj/coq/_build/default/theories/Numbers/Natural/Peano
Coq.Numbers.Natural.Binary
  /home/proj/coq/_build/default/theories/Numbers/Natural/Binary
Coq.Numbers.**Integer** /home/proj/coq/_build/default/theories/Numbers/Integer
Coq.Numbers.Integer.NatPairs
  /home/proj/coq/_build/default/theories/Numbers/Integer/NatPairs
Coq.Numbers.Integer.Abstract
  /home/proj/coq/_build/default/theories/Numbers/Integer/Abstract
Coq.Numbers.Integer.Binary
  /home/proj/coq/_build/default/theories/Numbers/Integer/Binary
Coq.Numbers.**NatInt** /home/proj/coq/_build/default/theories/Numbers/NatInt
Coq.Numbers.**Cyclic** /home/proj/coq/_build/default/theories/Numbers/Cyclic
Coq.Numbers.Cyclic.Int63
  /home/proj/coq/_build/default/theories/Numbers/Cyclic/Int63
Coq.Numbers.Cyclic.ZModulo
  /home/proj/coq/_build/default/theories/Numbers/Cyclic/ZModulo
Coq.Numbers.Cyclic.Int31
  /home/proj/coq/_build/default/theories/Numbers/Cyclic/Int31
Coq.Numbers.Cyclic.Abstract
  /home/proj/coq/_build/default/theories/Numbers/Cyclic/Abstract
Coq.**Arith** /home/proj/coq/_build/default/theories/Arith
Coq.**MSets** /home/proj/coq/_build/default/theories/MSets
Coq.**Floats** /home/proj/coq/_build/default/theories/Floats
Coq.**Unicode** /home/proj/coq/_build/default/theories/Unicode
Coq.**Wellfounded** /home/proj/coq/_build/default/theories/Wellfounded
Coq.**Classes** /home/proj/coq/_build/default/theories/Classes
Coq.**Lists** /home/proj/coq/_build/default/theories/Lists
Coq.**Strings** /home/proj/coq/_build/default/theories/Strings
Coq.**ZArith** /home/proj/coq/_build/default/theories/ZArith
Coq.**Reals** /home/proj/coq/_build/default/theories/Reals
Coq.Reals.Cauchy /home/proj/coq/_build/default/theories/Reals/Cauchy
Coq.Reals.Abstract /home/proj/coq/_build/default/theories/Reals/Abstract
Coq.**btauto** /home/proj/coq/_build/default/theories/btauto
Coq.**omega** /home/proj/coq/_build/default/theories/omega
Coq.**PArith** /home/proj/coq/_build/default/theories/PArith
Coq.**ssr** /home/proj/coq/_build/default/theories/ssr
Coq.**setoid_ring** /home/proj/coq/_build/default/theories/setoid_ring
Coq.**Sets** /home/proj/coq/_build/default/theories/Sets
Coq.**Init** /home/proj/coq/_build/default/theories/Init
Coq.**extraction** /home/proj/coq/_build/default/theories/extraction
Coq.**Setoids** /home/proj/coq/_build/default/theories/Setoids
Coq.**Vectors** /home/proj/coq/_build/default/theories/Vectors
Coq.**micromega** /home/proj/coq/_build/default/theories/micromega
Coq.**funind** /home/proj/coq/_build/default/theories/funind
Coq.**Structures** /home/proj/coq/_build/default/theories/Structures
Coq.**Array** /home/proj/coq/_build/default/theories/Array
Coq.**Sorting** /home/proj/coq/_build/default/theories/Sorting
Coq.**FSets** /home/proj/coq/_build/default/theories/FSets
Coq.**nsatz** /home/proj/coq/_build/default/theories/nsatz
Coq.**NArith** /home/proj/coq/_build/default/theories/NArith
Coq.**rtauto** /home/proj/coq/_build/default/theories/rtauto
Coq.**Relations** /home/proj/coq/_build/default/theories/Relations
Coq.**QArith** /home/proj/coq/_build/default/theories/QArith
Coq.**Logic** /home/proj/coq/_build/default/theories/Logic
Coq.**derive** /home/proj/coq/_build/default/theories/derive
Coq.**Compat** /home/proj/coq/_build/default/theories/Compat
<> /home/proj/coq

**Sorted**

Print LoadPath.

Logical Path / Physical path:
<> /home/jfehrle/.local/share/coq
<> /home/proj/coq/_build/default/user-contrib
Ltac2 /home/proj/coq/_build/default/user-contrib/Ltac2
Coq /home/proj/coq/_build/default/theories
Coq.Arith /home/proj/coq/_build/default/theories/Arith
Coq.Array /home/proj/coq/_build/default/theories/Array
Coq.Bool /home/proj/coq/_build/default/theories/Bool
Coq.Classes /home/proj/coq/_build/default/theories/Classes
Coq.Compat /home/proj/coq/_build/default/theories/Compat
Coq.FSets /home/proj/coq/_build/default/theories/FSets
Coq.Floats /home/proj/coq/_build/default/theories/Floats
Coq.Init /home/proj/coq/_build/default/theories/Init
Coq.Lists /home/proj/coq/_build/default/theories/Lists
Coq.Logic /home/proj/coq/_build/default/theories/Logic
Coq.MSets /home/proj/coq/_build/default/theories/MSets
Coq.NArith /home/proj/coq/_build/default/theories/NArith
Coq.Numbers /home/proj/coq/_build/default/theories/Numbers
Coq.Numbers.Cyclic
  /home/proj/coq/_build/default/theories/Numbers/Cyclic
Coq.Numbers.Cyclic.Abstract
  /home/proj/coq/_build/default/theories/Numbers/Cyclic/Abstract
Coq.Numbers.Cyclic.Int31
  /home/proj/coq/_build/default/theories/Numbers/Cyclic/Int31
Coq.Numbers.Cyclic.Int63
  /home/proj/coq/_build/default/theories/Numbers/Cyclic/Int63
Coq.Numbers.Cyclic.ZModulo
  /home/proj/coq/_build/default/theories/Numbers/Cyclic/ZModulo
Coq.Numbers.Integer
  /home/proj/coq/_build/default/theories/Numbers/Integer
Coq.Numbers.Integer.Abstract
  /home/proj/coq/_build/default/theories/Numbers/Integer/Abstract
Coq.Numbers.Integer.Binary
  /home/proj/coq/_build/default/theories/Numbers/Integer/Binary
Coq.Numbers.Integer.NatPairs
  /home/proj/coq/_build/default/theories/Numbers/Integer/NatPairs
Coq.Numbers.NatInt
  /home/proj/coq/_build/default/theories/Numbers/NatInt
Coq.Numbers.Natural
  /home/proj/coq/_build/default/theories/Numbers/Natural
Coq.Numbers.Natural.Abstract
  /home/proj/coq/_build/default/theories/Numbers/Natural/Abstract
Coq.Numbers.Natural.Binary
  /home/proj/coq/_build/default/theories/Numbers/Natural/Binary
Coq.Numbers.Natural.Peano
  /home/proj/coq/_build/default/theories/Numbers/Natural/Peano
Coq.PArith /home/proj/coq/_build/default/theories/PArith
Coq.Program /home/proj/coq/_build/default/theories/Program
Coq.QArith /home/proj/coq/_build/default/theories/QArith
Coq.Reals /home/proj/coq/_build/default/theories/Reals
Coq.Reals.Abstract
  /home/proj/coq/_build/default/theories/Reals/Abstract
Coq.Reals.Cauchy
  /home/proj/coq/_build/default/theories/Reals/Cauchy
Coq.Relations /home/proj/coq/_build/default/theories/Relations
Coq.Setoids /home/proj/coq/_build/default/theories/Setoids
Coq.Sets /home/proj/coq/_build/default/theories/Sets
Coq.Sorting /home/proj/coq/_build/default/theories/Sorting
Coq.Strings /home/proj/coq/_build/default/theories/Strings
Coq.Structures /home/proj/coq/_build/default/theories/Structures
Coq.Unicode /home/proj/coq/_build/default/theories/Unicode
Coq.Vectors /home/proj/coq/_build/default/theories/Vectors
Coq.Wellfounded
  /home/proj/coq/_build/default/theories/Wellfounded
Coq.ZArith /home/proj/coq/_build/default/theories/ZArith
Coq.btauto /home/proj/coq/_build/default/theories/btauto
Coq.derive /home/proj/coq/_build/default/theories/derive
Coq.extraction /home/proj/coq/_build/default/theories/extraction
Coq.funind /home/proj/coq/_build/default/theories/funind
Coq.micromega /home/proj/coq/_build/default/theories/micromega
Coq.nsatz /home/proj/coq/_build/default/theories/nsatz
Coq.omega /home/proj/coq/_build/default/theories/omega
Coq.rtauto /home/proj/coq/_build/default/theories/rtauto
Coq.setoid_ring
  /home/proj/coq/_build/default/theories/setoid_ring
Coq.ssr /home/proj/coq/_build/default/theories/ssr
Coq.ssrmatching
  /home/proj/coq/_build/default/theories/ssrmatching
<> /home/proj/coq
</details>

### CODE REVIEW

**Comment by jfehrle** [comment]
Failing vst job was killed, fiat_ crypto got an interrupt.  Probably unrelated.

**Comment by Zimmi48** [comment]
> I think it's unlikely (perhaps impossible) this reordering will impact users. Given that the order is currently nondeterministic, this seems like a reasonable thing to do. We'll see if CI makes me a liar.

It's probably impossible for this change to affect users since #14718:

>it is now an error to require a file using a partial logical name which does not resolve to a non-ambiguous path

> Failing vst job was killed, fiat_ crypto got an interrupt. Probably unrelated.

Yes, this happens all the time. These are OOM on the new runners. See [vst failing](https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs-.26-plugin-devs/topic/vst.20failing), [fiat crypto broken](https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs-.26-plugin-devs/topic/fiat.20crypto.20broken) and [Runner failures on ci-coq-02-runner-*](https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs-.26-plugin-devs/topic/Runner.20failures.20on.20ci-coq-02-runner-*).

**Comment by ejgallego** [comment]
I have no idea about this, what are the best practices in the C world regading sorting readdir ?

**Comment by jfehrle** [comment]
Well, `ls` sorts the entries by default, but this behavior is not specific to C nor really a C-related issue.  It applies to any language that reads a directory.  From a user interface perspective, sorting items is best practice when the original order is meaningless.  As in this case.

**Comment by ejgallego** [comment]
> Well, `ls` sorts the entries by default, but this behavior is not specific to C nor really a C-related issue. It applies to any language that reads a directory. From a user interface perspective, sorting items is best practice when the original order is meaningless. As in this case.

Then, if that's a user interface issue, shouldn't be the user interface the one sorting out the entries?

Note that this is a low-level function in the system module. I'm asking about C because indeed it is the standard language for the Unix API, but indeed that's a Unix best-practice for `readdir` what I'm interested about.

**Comment by jfehrle** [comment]
It would be considerably more complex to do it in `Print LoadPath` command, and it would be a bit misleading to do so only in the command.  I have trouble seeing why a benign change in System is a problem just because it's helpful for the user experience.

What would you consider evidence of best practices beyond what `ls` does?  There are tons of things that make sense that no one has ever written up as a best practice.  And even then, how do you decide what sources you find credible/persuasive?

**Comment by jfehrle** [comment]
I can do it this way.  WDYT?

File: checker/checker.ml [hunk 1]
```diff
@@ -181,7 +181,7 @@ let print_usage_channel co command =
 \n  -Q dir coqdir               map physical dir to logical coqdir\
 \n  -R dir coqdir               synonymous for -Q\
 \n  -coqlib dir                 set coqchk's standard library location\
-\n  -boot                       don't initalize the library paths automatically\
+\n  -boot                       don't initialize the library paths automatically\
```
Comment (SkySkimmer) [inline checker/checker.ml:184]: Unrelated to PR title

Comment (jfehrle) [reply checker/checker.ml]: Picky, picky.  Changed it.

**Comment by jfehrle** [comment]
@coqbot: run full ci now

File: checker/checker.ml (hunk 1 above)
Comment (jfehrle) [reply checker/checker.ml]: @SkySkimmer Can you be the assignee?  Or suggest someone?  Thanks.

**Review by Zimmi48 — APPROVED** [review]
I think this is a reasonable change. If there is no opposition to it, I'll merge.

Maybe adding a changelog would still be a good idea.

**Comment by jfehrle** [comment]
I added a changelog.

**Comment by Zimmi48** [comment]
@coqbot run full CI

**Comment by jfehrle** [comment]
@Zimmi48 Is this still on your "to do" list?

**Comment by Zimmi48** [comment]
Yes, @coqbot merge now

**Pull request merged**